extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/logical/DruidLogicalToQueryDefinitionTranslator.java

```diff
@@ -0,0 +1,141 @@
+
+import org.apache.calcite.rel.RelNode;
+import org.apache.druid.error.DruidException;
+import org.apache.druid.java.util.common.Intervals;
+import org.apache.druid.msq.input.InputSpec;
+import org.apache.druid.msq.kernel.QueryDefinition;
+import org.apache.druid.msq.logical.LogicalStageBuilder.RootStage;
+import org.apache.druid.msq.querykit.DataSourcePlan;
+import org.apache.druid.query.DataSource;
+import org.apache.druid.query.InlineDataSource;
+import org.apache.druid.query.TableDataSource;
+import org.apache.druid.sql.calcite.planner.PlannerContext;
+import org.apache.druid.sql.calcite.planner.querygen.DruidQueryGenerator.DruidNodeStack;
+import org.apache.druid.sql.calcite.planner.querygen.SourceDescProducer.SourceDesc;
+import org.apache.druid.sql.calcite.rel.logical.DruidLogicalNode;
+import org.apache.druid.sql.calcite.rel.logical.DruidTableScan;
+import org.apache.druid.sql.calcite.rel.logical.DruidValues;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+public class DruidLogicalToQueryDefinitionTranslator
+{
+  private PlannerContext plannerContext;
+  private LogicalStageBuilder stageBuilder;
+
+  public DruidLogicalToQueryDefinitionTranslator(PlannerContext plannerContext)
+  {
+    this.plannerContext = plannerContext;
+    this.stageBuilder = new LogicalStageBuilder(plannerContext);
+  }
+
+  public QueryDefinition translate(DruidLogicalNode relRoot)
+  {
+    DruidNodeStack stack = new DruidNodeStack();
+    stack.push(relRoot);
+    LogicalStage logicalStage = buildStageFor(stack);
+    return logicalStage.build();
+  }
+
+  private LogicalStage buildStageFor(DruidNodeStack stack)
+  {
+    List<LogicalStage> newInputs = new ArrayList<>();
+
+    for (RelNode input : stack.peekNode().getInputs()) {
+      stack.push((DruidLogicalNode) input, newInputs.size());
+      newInputs.add(buildStageFor(stack));
+      stack.pop();
+    }
+    LogicalStage stage = processNodeWithInputs(stack, newInputs);
+    return stage;
+  }
+
+  private LogicalStage processNodeWithInputs(DruidNodeStack stack, List<LogicalStage> newInputs)
+  {
+    DruidLogicalNode node = stack.peekNode();
+    Optional<RootStage> stage = buildRootStage(node);
+    if (stage.isPresent()) {
+      return stage.get();
+    }
+    if (newInputs.size() == 1) {
+      LogicalStage inputStage = newInputs.get(0);
+      LogicalStage newStage = inputStage.extendWith(stack);
+      if (newStage != null) {
+        return newStage;
+      }
+      Optional<LogicalStage> seq = buildSequence(stack, newInputs.get(0));
+      if (seq.isPresent()) {
+        return seq.get();
+      }
+    }
+    throw DruidException.defensive().build("Unable to process relNode[%s]", node);
+  }
+
+  private Optional<LogicalStage> buildSequence(DruidNodeStack stack, LogicalStage stage)
+  {
+    return Optional.empty();
+  }
+
+  private Optional<RootStage> buildRootStage(DruidLogicalNode node)
```

Comment:
From the implementation, it looks like a "root stage" is meant to be a stage that has no other stages as input. This suggests a model where the edges in a DAG point from producer stage to consumer stage. However, in MSQ's conceptual model, edges point the other way, from consumer to producer. So stages with no other stages as input would be called "leaf stages" rather than "root stages". This choice of direction is arbitrary, but it would be good to be consistent.

Reply:
the usage of `Root` was misleading ; renamed it to `ReadStage`.